hazelcast-sql/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -0,0 +1,7271 @@
+                        addToSelectList(
+                                selectItems,
+                                aliases,
+                                fields,
+                                exp,
+                                scope,
+                                includeSystemVars);
+                    } else {
+                        final SqlNode from = SqlNonNullableAccessors.getNode(child);
+                        final SqlValidatorNamespace fromNs = getNamespaceOrThrow(from, scope);
+                        final RelDataType rowType = fromNs.getRowType();
+                        for (RelDataTypeField field : rowType.getFieldList()) {
+                            String columnName = field.getName();
+
+                            // TODO: do real implicit collation here
+                            final SqlIdentifier exp =
+                                    new SqlIdentifier(
+                                            ImmutableList.of(child.name, columnName),
+                                            startPosition);
+                            // Don't add expanded rolled up columns
+                            if (!isRolledUpColumn(exp, scope)) {
+                                addOrExpandField(
+                                        selectItems,
+                                        aliases,
+                                        fields,
+                                        includeSystemVars,
+                                        scope,
+                                        exp,
+                                        field);
+                            }
+                        }
+                    }
+                    if (child.nullable) {
+                        for (int i = before; i < fields.size(); i++) {
+                            final Map.Entry<String, RelDataType> entry = fields.get(i);
+                            final RelDataType type = entry.getValue();
+                            if (!type.isNullable()) {
+                                fields.set(i,
+                                        Pair.of(entry.getKey(),
+                                                typeFactory.createTypeWithNullability(type, true)));
+                            }
+                        }
+                    }
+                }
+                // Commenting out this section:
+                // https://github.com/hazelcast/hazelcast/issues/21325
+                //
+                // If NATURAL JOIN or USING is present, move key fields to the front of
+                // the list, per standard SQL. Disabled if there are dynamic fields.
+                // if (!hasDynamicStruct || Bug.CALCITE_2400_FIXED) {
+                //    SqlNode from = requireNonNull(scope.getNode().getFrom(),
+                //            () -> "getFrom for " + scope.getNode());
+                //    new Permute(from, 0).permute(selectItems, fields);
+                // }
```

Comment:
With this you essentially broke this type of join. For example, if you have `t1(a, b)` and `t2(c, b)` and you do `FROM t1 JOIN t2 USING (b)` or `FROM t1 NATURAL JOIN t2`, the query is supposed to have these columns on the output: `b, a, c`. With this code commented out, it has `a, b, c, b` - `b` is there twice with the same value, and it should be at the beginning, as the calcite code comment says